ADBDEV-3656: Vacuum make all segments inaccessible in case of concurrent transaction present

RekGRpth opened the pull request:
Fix unavailability of all segment files after vacuum

After compaction the vacuum query always marked the segment files as awaiting
drop in the coordinator's hashmap, regardless of their actual state on segments.
Because the compaction might not have happened, the segment files might have
remained in their normal state, but coordinator did not get information about
this. Drop phase may be skipped in the case of there are concurrent
transactions, so segment files were left in awaiting drop state in
the coordinator's hashmap. Therefore on intensive concurrent insertion into
a table, vacuum marked all available segment files as awaiting drop in
the coordinator's hashmap and the next insert failed with message:

```sql
ERROR: could not find segment file to use for inserting into relation
```

After the patch the coordinator extracts information from the segments about
the actual state of the segment files after compaction and marks as awaiting
drop only those segment files that are actually compacted. This allows for
intensive concurrent insertion into the table.

Stolb27 commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -778,6 +778,54 @@ RegisterSegnoForCompactionDrop(Oid relid, List *compactedSegmentFileList)
+	}
+
+	aosegrel = relation_open(relid, AccessShareLock);
+	awaiting_drop = get_awaiting_drop_status_from_segments(aosegrel);
```
There are dispatch results for each vacuum operation. Did you research possibility to use/extend them?

RekGRpth replied:
> There are dispatch results for each vacuum operation. Did you research possibility to use/extend them?

Not yet, but it seems that information about the state of the segment file can be transmitted through them.

RekGRpth replied:
I tried to do this, but it turned out somehow too monstrous: https://github.com/arenadata/gpdb/pull/549

andr-sokolov replied:
@RekGRpth, what about rewriting GetTotalTupleCountFromSegments()? 
Its call stack is:

```
GetTotalTupleCountFromSegments() at appendonlywriter.c:1375
UpdateMasterAosegTotalsFromSegments() at appendonlywriter.c:1626
vacuum_rel () at vacuum.c:2579
vacuum_rel_ao_phase () at vacuum.c:2252
vacuumStatement_Relation () at vacuum.c:890
vacuum () at vacuum.c:353
```

The GetTotalTupleCountFromSegments() function executes "`SELECT tupcount, segno FROM pg_aoseg.pg_aoseg_16384`" and
the get_awaiting_drop_status_from_segments() function executes "`SELECT state, segno FROM pg_aoseg.pg_aoseg_16384`".
Can we execute one query "`SELECT tupcount, state, segno FROM pg_aoseg.pg_aoseg_16384`" instead of two queries?

RekGRpth replied:
> what about rewriting GetTotalTupleCountFromSegments()? Its call stack is:
> 
> ```
> GetTotalTupleCountFromSegments() at appendonlywriter.c:1375
> UpdateMasterAosegTotalsFromSegments() at appendonlywriter.c:1626
> vacuum_rel () at vacuum.c:2579
> vacuum_rel_ao_phase () at vacuum.c:2252
> vacuumStatement_Relation () at vacuum.c:890
> vacuum () at vacuum.c:353
> ```
> 
> The GetTotalTupleCountFromSegments() function executes "`SELECT tupcount, segno FROM pg_aoseg.pg_aoseg_16384`" and the get_awaiting_drop_status_from_segments() function executes "`SELECT state, segno FROM pg_aoseg.pg_aoseg_16384`". Can we execute one query "`SELECT tupcount, state, segno FROM pg_aoseg.pg_aoseg_16384`" instead of two queries?

Yes, you are right, this also works, and it looks more optimal. See https://github.com/arenadata/gpdb/pull/562

RekGRpth replied:
done

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -778,6 +778,54 @@ RegisterSegnoForCompactionDrop(Oid relid, List *compactedSegmentFileList)
+
+	release_lightweight_lock();
+	pfree(awaiting_drop);
+	return;
```
return at the end of function is redundant

RekGRpth replied:
Yes, but see line https://github.com/arenadata/gpdb/pull/548/files#diff-9fecdbbedbe0a8ed6898b84ef38c28e23ae4a8f4d6f09b38f57be565c4ee8504R778

andr-sokolov replied:
I don't know what the author means. May be he thinks that a function with return statement at the end and the same function without it work differently. But I don't know the difference. So I suggest to remove redundant return

RekGRpth replied:
fixed

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -778,6 +778,53 @@ RegisterSegnoForCompactionDrop(Oid relid, List *compactedSegmentFileList)
+	aoentry = AORelGetOrCreateHashEntry(relid);
+	Assert(aoentry);
+
+	for (i = 0; i < MAX_AOREL_CONCURRENCY; i++)
```
Is it possible to use foreach over compactedSegmentFileList instead of this loop? It seems that it will be faster

RekGRpth replied:
I just copied RegisterSegnoForCompactionDrop and modified a bit.

RekGRpth replied:
fixed

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -778,6 +778,53 @@ RegisterSegnoForCompactionDrop(Oid relid, List *compactedSegmentFileList)
+
+	for (i = 0; i < MAX_AOREL_CONCURRENCY; i++)
+	{
+		AOSegfileStatus *segfilestat = &aoentry->relsegfiles[i];
```
The segfilestat variable is used only one time or never if condition below is false. I suggest to remove this variable and write "aoentry->relsegfiles[i].state = DROP_USE"

RekGRpth replied:
I just copied RegisterSegnoForCompactionDrop and modified a bit.

andr-sokolov replied:
In RegisterSegnoForCompactionDrop the segfilestat variable is used two times. We have no goal to write a function which will be as similar as possible to RegisterSegnoForCompactionDrop

RekGRpth replied:
fixed

andr-sokolov commented on src/test/isolation2/input/uao/vacuum_while_max_concurrency.source:
```diff
@@ -0,0 +1,407 @@
+SELECT * FROM ao;
+SELECT * FROM gp_ao_or_aocs_seg('ao') ORDER BY segno;
+
+ALTER RESOURCE GROUP admin_group SET CONCURRENCY 20;
```
The default value for admin_group is 10, not 20. I suggest to put this query before the first ALTER to be sure that the second ALTER reset the limit correctly
```
SELECT c.value
FROM pg_authid a
  JOIN pg_resgroupcapability c ON c.resgroupid = a.rolresgroup
  JOIN pg_resourcetype t ON t.restypid = c.reslimittype
WHERE a.rolname = current_user
  AND t.resname = 'active_statements';
```

RekGRpth replied:
I just copied from https://github.com/arenadata/gpdb/blob/6b9f47498afda1b75e307a60a965b40ff594ff57/src/test/isolation2/input/uao/max_concurrency.source#L392

RekGRpth replied:
That is, the previous test has already overwritten the default value.

andr-sokolov replied:
Ok, let it reset to 20, but the SELECT is still needed to be sure that some other test doesn't change CONCURRENCY to another value

RekGRpth replied:
If I hardcode the check for 20 in the test, it will fail on a single run.

RekGRpth replied:
https://github.com/arenadata/gpdb/blob/6b9f47498afda1b75e307a60a965b40ff594ff57/gpdb-doc/markdown/ref_guide/sql_commands/CREATE_RESOURCE_GROUP.html.md?plain=1#L48

RekGRpth replied:
I removed the resource group change.

andr-sokolov commented on src/test/isolation2/input/uao/vacuum_while_max_concurrency.source:
```diff
@@ -0,0 +1,407 @@
+DROP TABLE IF EXISTS ao;
+CREATE TABLE ao (a INT, b INT) WITH (appendonly=true, orientation=@orientation@);
+
+ALTER RESOURCE GROUP admin_group SET CONCURRENCY 130;
```
Write please comment before ALTER how 130 was calculated. 127 - for parallel sessions. What are the other 3 used for?

RekGRpth replied:
I removed the resource group change.

andr-sokolov commented on src/test/isolation2/output/uao/vacuum_while_max_concurrency.source:
```diff
@@ -0,0 +1,1063 @@
+(131 rows)
+SELECT * FROM gp_ao_or_aocs_seg('ao') ORDER BY segno;
+ segno | tupcount | modcount | formatversion | state 
+-------+----------+----------+---------------+-------
```
Here we should check that the state of all file segments is AVAILABLE (1). Values in the tupcount, modcount and formatversion columns do not matter. Let's remove these columns from the output

RekGRpth replied:
fixed

andr-sokolov commented on src/test/isolation2/output/uao/vacuum_while_max_concurrency.source:
```diff
@@ -0,0 +1,1063 @@
+5: COMMIT;
+COMMIT
+
+SELECT * FROM ao;
```
The select query outputs tuples in arbitrary order. I suggest to add `ORDER BY a,b` to be sure that the order will always be the same. Do we really need this SELECT? What does it check?

RekGRpth replied:
fixed

andr-sokolov commented on src/test/isolation2/input/uao/vacuum_while_max_concurrency.source:
```diff
@@ -0,0 +1,403 @@
+-- @Description Insert into a ao relation with 127 concurrent transactions
```
The description is invalid. The test checks that file segments are available for insert after vacuum in the case of parallel transaction. Provide please actual test description here and comment each step of the test. Describe in the comments why the test opens 127 parallel sessions, why the test does 4 parallel inserts, what is checked with SELECTs, how the numbers was calculated (why not 120 and not 5).

RekGRpth replied:
fixed

andr-sokolov commented on src/test/isolation2/input/uao/vacuum_while_max_concurrency.source:
```diff
@@ -0,0 +1,403 @@
+127: COMMIT;
+
+1: BEGIN;
+1: INSERT INTO ao SELECT 2, 1;
```
I suggest to add checks for number of available file segments to this block after each INSERT and VACUUM

RekGRpth replied:
fixed

andr-sokolov commented on src/test/isolation2/input/uao/vacuum_while_max_concurrency.source:
```diff
@@ -0,0 +1,403 @@
+-- @Description Insert into a ao relation with 127 concurrent transactions
+-- 
+DROP TABLE IF EXISTS ao;
```
The output of this query differs in the case of table exists andin the case of table doesn't exist. But it does not matter for the test. Let's ignore the query output.

RekGRpth replied:
done

InnerLife0 commented:
@RekGRpth 
Every time you push a single commit or left a single comment, it produces each own email for everyone who subscribed to your PR.
![image](https://github.com/arenadata/gpdb/assets/42867699/b46ae56d-809d-4778-9eeb-7e56d764642b)
![image](https://github.com/arenadata/gpdb/assets/42867699/5f6dd200-a4e3-433e-bd53-30a03335d1ae)

I have an inbox full of meaningless messages from Georgy Shelkovy, which looks like a spam. All of them can be collapsed to two messages: One push for all commits and one as reply to all messages.

andr-sokolov commented:
I suggest to replace the commit and PR description with:

> Fix unavailability of all segment files after vacuum
> 
> After compaction the vacuum query always marked the segment files as awaiting
> drop in the coordinator's hashmap, regardless of their actual state on segments.
> Because the compaction might not have happened, the segment files might have
> remained in their normal state, but coordinator did not get information about
> this. Drop phase may be skipped in the case of there are concurrent
> transactions, so segment files were left in awaiting drop state in
> the coordinator's hashmap. Therefore on intensive concurrent insertion into
> a table, vacuum marked all available segment files as awaiting drop in
> the coordinator's hashmap and the next insert failed with message:
> 
> ERROR: could not find segment file to use for inserting into relation
> 
> After the patch the coordinator extracts information from the segments about
> the actual state of the segment files after compaction and marks as awaiting
> drop only those segment files that are actually compacted. This allows for
> intensive concurrent insertion into the table.

Major differences:
1. The "fix" word in the title point that the patch fixed the error, it is not a new feature
2. The title is less than 80 symbols
3. Past tense is used to tell about the error, present tense is used to tell about right behaviour. Now it is easier to understand what the patch changes and what does not.
4. I have reformulated sentences to eliminate the use of parentheses and strange collocation like "patch extracts information"

RekGRpth commented:
> I suggest to replace the commit and PR description with:
> 
> > Fix unavailability of all segment files after vacuum
> > After compaction the vacuum query always marked the segment files as awaiting
> > drop in the coordinator's hashmap, regardless of their actual state on segments.
> > Because the compaction might not have happened, the segment files might have
> > remained in their normal state, but coordinator did not get information about
> > this. Drop phase may be skipped in the case of there are concurrent
> > transactions, so segment files were left in awaiting drop state in
> > the coordinator's hashmap. Therefore on intensive concurrent insertion into
> > a table, vacuum marked all available segment files as awaiting drop in
> > the coordinator's hashmap and the next insert failed with message:
> > ERROR: could not find segment file to use for inserting into relation
> > After the patch the coordinator extracts information from the segments about
> > the actual state of the segment files after compaction and marks as awaiting
> > drop only those segment files that are actually compacted. This allows for
> > intensive concurrent insertion into the table.
> 
> Major differences:
> 
>     1. The "fix" word in the title point that the patch fixed the error, it is not a new feature
> 
>     2. The title is less than 80 symbols
> 
>     3. Past tense is used to tell about the error, present tense is used to tell about right behaviour. Now it is easier to understand what the patch changes and what does not.
> 
>     4. I have reformulated sentences to eliminate the use of parentheses and strange collocations like "patch extracts information"

done

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -1324,7 +1325,8 @@ assignPerRelSegno(List *all_relids)
  */
 static int64 *
 GetTotalTupleCountFromSegments(Relation parentrel,
```
The function has been changed, so its description should be changed too.

RekGRpth replied:
done

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -1402,6 +1405,31 @@ GetTotalTupleCountFromSegments(Relation parentrel,
+						value = PQgetvalue(pgresult, j, 2);
+						qe_state = pg_atoi(value, sizeof(int32), 0);
+
+						ValidateAppendonlySegmentDataBeforeStorage(segno);
```
This check should be added before the first usage of segno (`total_tupcount[segno] += tupcount;`)

RekGRpth replied:
done

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -1641,6 +1674,37 @@ UpdateMasterAosegTotalsFromSegments(Relation parentrel,
+			{
+				ereportif(Debug_appendonly_print_segfile_choice, LOG,
+						(errmsg("Update segno %d after compaction "
+								"relation \"%s\" (%d)", qe_segno,
```
I suggest to rewrite the message, because "after compaction relation" is a strange collocation. For example, "Update the file segment state after compaction: relation \"%s\" (%d), segno %d"

RekGRpth replied:
done

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -1561,22 +1596,29 @@ get_awaiting_drop_status_from_segments(Relation parentrel)
 /*
  * Updates the tupcount information from the segments.
+ * If list as fifth argument is given then updates state of segfiles,
+ * which is not really compacted on all segments, in aoentry hashmap.
```
segfiles -> which ARE not really

RekGRpth replied:
> segfiles -> which ARE not really

fixed

Stolb27 commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -1641,6 +1683,38 @@ UpdateMasterAosegTotalsFromSegments(Relation parentrel,
+					   RelationGetRelid(parentrel),
+					   qe_segno);
+
+				aoentry->relsegfiles[qe_segno].state = DROP_USE;
```
@RekGRpth, @andr-sokolov, @dnskvlnk, WHY is segment marked as DROP_USE, if compaction didn't happen? Are there any cases when this state will stay here and data will be unavailable?

RekGRpth replied:
> WHY is segment marked as DROP_USE

Because the state machine at the end of the transaction will scroll the state to the default value.

Stolb27 replied:
> Because the state machine at the end of the transaction will scroll the state to the default value.

in case of commit only. What about abort? Anyway, it's very dirty hack and may be unexpected for further developers.

RekGRpth replied:
I changed to use INSERT_USE, then the state machine will scroll the state to the default value regardless of the commit or abort.

RekGRpth replied:
Or then we need to enter one more state "NOT_COMPACTED" and process all transitions correctly.

Stolb27 replied:
Can you explain, why INSERT_USE? After your modifications, vacuum doesn't touch given AO-segment at all. What about sideeffects of this state?

RekGRpth replied:
I changed it to the default value and added its processing to the state machine.

Stolb27 replied:
Why do we need additional actions for default state? Do we really need them?

Moreover, you MUST describe each of your changes to state machine description.

RekGRpth replied:
Yes, you are right, I changed it.

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -1641,6 +1683,41 @@ UpdateMasterAosegTotalsFromSegments(Relation parentrel,
+					   qe_segno);
+
+				segfilestat->state = AVAILABLE;
+				segfilestat->xid = InvalidTransactionId;
```
Is it possible that aoentry->txns_using_rel will not be decremented in AtEOXact_AppendOnly_Relation() because xid is InvalidTransactionId?

RekGRpth replied:
fixed

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -1641,6 +1683,40 @@ UpdateMasterAosegTotalsFromSegments(Relation parentrel,
+		{
+			int			qe_segno = lfirst_int(l);
+
+			Assert(qe_segno >= 0);
```
I suggest
```
Assert(qe_segno >= 0 && qe_segno < MAX_AOREL_CONCURRENCY); 
```

RekGRpth replied:
fixed

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -1641,6 +1683,40 @@ UpdateMasterAosegTotalsFromSegments(Relation parentrel,
+					   RelationGetRelid(parentrel),
+					   qe_segno);
+
+				segfilestat->state = AVAILABLE;
```
Let's remove the redundant segfilestat variable:
```
aoentry->relsegfiles[qe_segno].state = AVAILABLE;
```

RekGRpth replied:
done

andr-sokolov commented on src/backend/access/appendonly/appendonlywriter.c:
```diff
@@ -1641,6 +1683,40 @@ UpdateMasterAosegTotalsFromSegments(Relation parentrel,
+
+		foreach(l, appendonly_compaction_segno)
+		{
+			int			qe_segno = lfirst_int(l);
```
Why is the variable name qe_segno? What does "qe" mean?

RekGRpth replied:
renamed

RekGRpth commented:
The vacuum always selects one segment file to compact and sets the current transaction ID to the corresponding hashmap entry. https://github.com/arenadata/gpdb/blob/e655547faf9ff63aa4184bd9b344ffb6347ac090/src/backend/access/appendonly/appendonlywriter.c#L1067-L1069

RekGRpth commented:
Therefore, when the vacuum ends, there will always be one entry in the hashmap with the current transaction ID. https://github.com/arenadata/gpdb/blob/e655547faf9ff63aa4184bd9b344ffb6347ac090/src/backend/access/appendonly/appendonlywriter.c#L2013-L2020

RekGRpth committed:
Fix unavailability of all segment files after vacuum

After compaction the vacuum query always marked the segment files as awaiting
drop in the coordinator's hashmap, regardless of their actual state on segments.
Because the compaction might not have happened, the segment files might have
remained in their normal state, but coordinator did not get information about
this. Drop phase may be skipped in the case of there are concurrent
transactions, so segment files were left in awaiting drop state in
the coordinator's hashmap. Therefore on intensive concurrent insertion into
a table, vacuum marked all available segment files as awaiting drop in
the coordinator's hashmap and the next insert failed with message:

ERROR: could not find segment file to use for inserting into relation

After the patch the coordinator extracts information from the segments about
the actual state of the segment files after compaction and marks as awaiting
drop only those segment files that are actually compacted. This allows for
intensive concurrent insertion into the table.

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3656-4

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3656-4

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52597
